File: Data Structures & Algorithms/asteroid-collision/submission-2.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = [asteroids[0]]

        def isPositive(n):
            if n > 0:
                return True
            else:
                return False

        for a in asteroids[1:]:
            while True:
                if (not stack) or (isPositive(stack[-1]) == isPositive(a)) or (not isPositive(stack[-1])):
                    stack.append(a)
                    break
                else:
                    l = abs(stack[-1])
                    r = abs(a)

                    if l < r:
                        stack.pop()
                    elif l == r:
                        stack.pop()
                        break
                    else:
                        break
        
        return stack

File: Data Structures & Algorithms/valid-tree/submission-2.cpp
class Solution {
public:
    bool validTree(int n, vector<vector<int>>& edges) {
        vector<vector<int>> nodes(n, vector<int>{});
        vector<bool> visited(n, false);
        stack<pair<int, int>> st;
        
        for (auto& edge : edges) {
            nodes[edge[0]].push_back(edge[1]);
            nodes[edge[1]].push_back(edge[0]);
        }

        visited[0] = true;
        st.push({0, -1});
        int count = 0;

        while (!st.empty()) {
            pair<int, int> node = st.top();
            //cout << node.first << endl;
            st.pop();
            count++;

            for (auto& child : nodes[node.first]) {
                //cout << "child: " << child << endl;
                if (child != node.second) {
                    if (visited[child]) return false;

                    st.push({child, node.first});
                    //cout << "pushing child " << child << " with parent " << node.first << endl;
                    visited[child] = true;
                }
            }
        }

        //cout << "count: " << count << endl;

        bool retVal = count == n ? true : false;

        return retVal;
    }
};
